8338686: App classpath mismatch if a jar from the Class-Path attribute is on the classpath

calvinccheung opened the pull request:
When a CDS archive is created with `-cp A.jar:B.jar` where the manifest of A.jar contains a class-path attribute (e.g. `Class-Path: a.jar`). The following paths will be stored in the archive: A.jar:a.jar::B.jar.

Currently, if the user uses -cp A.jar:a.jar:B:jar during dump time and runtime, during runtime it will result in class path mismatch and the archive will not be used. It is due to during dump time when the ClassPathEntry's are being created, it checks if the name is in one of the existing entries and skip creating it. A simple fix is to add the check if the new entry and the existing entry are from the class path attribute.

With the fix, the following class path entries will be created during dump time:
A.jar:a.jar:a.jar:B.jar
The first a.jar is from the class path attribute, the second one is from the -cp. During runtime, the entries from the class path attribute obtained from the CDS archive will be skipped.

Testing: tiers 1 - 5.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338686](https://bugs.openjdk.org/browse/JDK-8338686): App classpath mismatch if a jar from the Class-Path attribute is on the classpath (**Bug** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20697/head:pull/20697` \
`$ git checkout pull/20697`

Update a local copy of the PR: \
`$ git checkout pull/20697` \
`$ git pull https://git.openjdk.org/jdk.git pull/20697/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20697`

View PR using the GUI difftool: \
`$ git pr show -t 20697`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20697.diff">https://git.openjdk.org/jdk/pull/20697.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20697#issuecomment-2307883061)

dholmes-ora commented on test/hotspot/jtreg/runtime/cds/appcds/ClassPathAttr.java:
```diff
@@ -126,6 +128,52 @@ static void testNonExistentJars() throws Exception {
+
+    // Run with the same -cp apattr1.jar:hello.jar. The Hello class should be
+    // loaded from the archive.
+    TestCommon.run(
+        "-Xlog:class+path,class+load",
+        "-cp", cp,
+        "Hello")
+      .assertNormalExit(output -> {
+          output.shouldContain("Hello source: shared objects file");
+        });
```
Style nit here and elsewhere:
```suggestion
    TestCommon.run("-Xlog:class+path,class+load",
                   "-cp", cp,
                   "Hello")
               .assertNormalExit(output -> {
                     output.shouldContain("Hello source: shared objects file");
                 });
```

dholmes-ora reviewed (commented):
Seems reasonable. A few style nits.

Thanks

calvinccheung commented:
> Seems reasonable. A few style nits.
> 
Hi David,
Thanks for your review. After discussion with Ioi, we have a much simpler fix. In the new fix, the change is only in classloader.cpp. I've fixed the nits in the test case and added another test scenario. The other nits are no longer applicable.
Thanks!

dholmes-ora reviewed (approved):
Seems reasonable but I confess I'm unclear what the actual rules should be here.

Thanks

iklam reviewed (approved):
LGTM

calvinccheung commented:
Thanks @dholmes-ora, @iklam for the review!

/integrate
